### ISSUE TITLE
[selectors] decide on :blank

### ISSUE THREAD

**Issue by annevk** [w3c/csswg-drafts#1967]
Is `:whitespace-only` really that bad? It's a more niche case so longer names seem acceptable then? Or just decide on `:blank` and remove the issue marker, allowing implementations to remove the prefixed variant (if still feasible).

**Comment by js-choi** [comment]
Is there a particular reason why `:ws-only` would be unacceptable? It contains what would be a new abbreviation in CSS, but it is much pithier while remaining semantically ambiguous, from what I can tell.

In addition, though perhaps less importantly, `:whitespace-only` is inconsistent with the already-introduced `white-space` CSS property, which considers “white space” to be a two-word phrase.

(My apologies if this particular option has already been discussed; I could find no mention of “ws-only” in this GitHub repository, the www-style archives, or the web in general. I can only imagine the tedious bikeshedding that has been occurring for the past four-to-five years.)

**Comment by fantasai** [comment]
Fwiw, there was also a discussion around maybe redefining `:empty` to allow white space, since that's what most people would expect it to be. The main question there would be Web-compat, I guess.

**Comment by bradkemper** [comment]
Yeah, the white space only version would be far less niche than `:empty` is. 

My preference is to make `:empty` into a functional notation, and if you leave off the parentheses it is the same as `:empty(legacy)` or whatever. Then you could have `:empty()` be the more useful version that ignores white space, and have other possible arguments to say whether non-breaking spaces, thin spaces, etc. are significant.

**Comment by whileloopa** [comment]
I suggest
:empty(strict) -- match elements with nothing inside
:empty() -- match elements that contain only '\s', '\t' and '\n'
:empty(blank) -- match elements that contain  '\s', '\t', '\n' and descendant elements that produce only whitespace, eg. <p></p>, <br/>, <pre>  </pre>, etc...

**Comment by annevk** [comment]
To be clear, I'm adding Agenda+ back since while it was discussed, no conclusion was reached and it's nearly a year later now without a path forward for implementations.

**Comment by fantasai** [comment]
Related issue: https://github.com/w3c/csswg-drafts/issues/1283

**Comment by AmeliaBR** [comment]
I really don't like the idea of a breaking change!

An example of a use-case this change would break: using `:empty` on a `<div contenteditable></div>` to set a style indicating that it hasn't been edited or show a placeholder message indicating that editing is possible ([old demo of mine](https://codepen.io/AmeliaBR/details/QvMGmR)). It would be rather confusing if the author did edit the content but the placeholder didn't disappear.

**Comment by kizu** [comment]
@AmeliaBR thinking more about it: while I agree that this exact use case would “break”, for this use case the effect would be still minimal and acceptable, in my opinion. While not disappeared placeholder on whitespace input would be slightly confusing, it won't break anything — the field would work the same, the only slight annoyance would happen only if people would start writing a message from a whitespace, and the only case where this could be highly confusing is if the main/only purpose of the input would be to contain just whitespace, but the intersection of this exact use case and an `:empty` usage could be non-existent?

So, while I generally prefer things to be backward-compatible, in this case, I don't see enough realistic use cases where things would break considerably, and the benefits of making the default behaviour more useful and less confusing would prevail over the minor annoyances it could cause.

**Comment by nigelmegitt** [comment]
_Extreme edge case alert_: an editor for the [Whitespace](https://en.wikipedia.org/wiki/Whitespace_(programming_language)) language might be broken by this change.

**Comment by kizu** [comment]
Only if it would use the `:empty` in this way.

Also: there is a chance that for those writing a Whitespace editor having the new default behavior for `:empty` could be potentially beneficial? They could use it as a very cheap validation tool, as you could target an editor containing something other than whitespaces with `:not(:empty)` :)

**Comment by fantasai** [comment]
@AmeliaBR I think that such a niche use case is far outweighed by the general confusion of :empty not selecting elements which the author perceives as being empty, for example P, LI, TD, and TH elements without end tags in a properly-indented source document. I think one of the major use cases for :empty is likely to be selecting empty cells in particular, and matching the conditions that the 'empty-cells' property uses seems likely to be much more useful and intuitive for authors than considering white space to be significant content here.

If there are strong use cases for other definitions of emptiness in a selector, we can consider adding them through a functional notation in the future, as @bradkemper suggests.

**Comment by fantasai** [comment]
@annevk Let me know if this is resolved to your satisfaction, else re-open?

**Comment by annevk** [comment]
I'm a little concerned that not many implementers weighed in on that decision. Are implementation bugs filed? Are the tests updated?

Also, your commit mutates the changelog in a way that's confusing me. As `:blank` is still there (though with new semantics).

**Comment by myakura** [comment]
> Also, your commit mutates the changelog in a way that's confusing me. As `:blank` is still there (though with new semantics).

@annevk that confused me too. but it appears the `:blank` introduced in 2b42b64a79dad888b10316e79675e2b507883eed is a brand new selector to target empty form elements. see #1283 and  https://www.w3.org/blog/CSS/2018/09/27/minutes-2018-09-26/

**Comment by ExE-Boss** [comment]
> Are implementation bugs filed?

Firefox has [bug&nbsp;1106296](https://bugzil.la/1106296).

**Comment by annevk** [comment]
Thanks for linking that. In that bug @dbaron wrote the following 3 years ago:

> Today the group resolved to drop :blank and change :empty to be like :blank, although I'm skeptical that this will be Web-compatible.  We should do this after Blink does it, but I don't think it's worth risking Web compat on before they move first.

It seems the CSS WG decided to try this again 3 years later unless I'm missing something? Why are we expecting different results? Why didn't it happen last time?

**Comment by bradkemper** [comment]
Is there a definition of what is considered white space for this issue? Is it just spaces, tabs, and line breaks/feeds? I would really like a version that also ignored nonbreaking spaces, thin spaces, etc. 

Even if those were to be added later as arguments to a function, making the new version be `:empty()` with the parentheses would prevent backward compatibility problems, while also setting it up for accepting arguments in the future.

**Comment by fantasai** [comment]
@bradkemper Yes, it's just spaces, tabs, and line breaks -- things that are used for formatting source code and can be collapsed away. Otherwise, you should just not put those spaces into the document. :)

**Comment by ExE-Boss** [comment]
I am personally in favour of how `:blank` and `:empty` are defined right now in commit https://github.com/w3c/csswg-drafts/blob/5f2e1b53e79b2c598ced822b80573202452e575e/selectors-4/Overview.bs.

**Comment by Dan503** [comment]
I need some clarification on this. 

So if `:blank` is being dropped completely, is `:empty` being updated to also cover `<input>` and `<textarea>` elements that have no text in them?

If yes, does white space count as content in these form elements?
I agree completely that for regular HTML elements `:empty` should not count white space as content. For form elements though I think it should count.

The primary use case for supporting `:empty` on form elements is for supporting the floating label design pattern.

```html
<label>
    <input/>
    <span>label text</span>
</label>
```

```css
input:focus + span,
input:empty + span {
    /* styles that move the label out of the input area */
}
```

When the input is empty and doesn't have focus, the label looks like placeholder text. When it has focus, or has content, the label moves out of the way.

I think it is useful for users if, for form elements, `:empty` counts white space as content. That way, if the user has entered spaces and no text, and the user has removed focus from the field, the label still displays in label form rather than placeholder form.

The advantage to the user here is that the field looks empty but since the label is displaying differently, the user is able to recognise that there is some white space content in there. If this matters for form submission then it is better for the user to be aware of what fields have only white space in them. Having white space only form submissions probably isn't that big of a deal though so having `:empty` not count white space in form elements isn't that bad.

**Comment by ExE-Boss** [comment]
I’d rather have `:blank` for user input elements, since `<textarea>` uses `textContent` as the textual content, and as such, `:empty` would have to behave differently for it, which results in a selector that ignores whitespace except for some elements, which goes against the goal of simplicity.

**Comment by Dan503** [comment]
I would be in favor of `:blank` being exclusively for form elements and counts white space as content and `:empty` being exclusively for HTML elements and not counting white space as content.

**Comment by zocky** [comment]
It's obvious that `:empty` should be left as is, and a new pseudo should be used for elements that have no other content than white space. `:blank` is as good a choice as any.

Also, inputs should be entirely left out of this. It is coincidental that this can be used to detect empty textareas. This cannot be made useful for detecting values of inputs in general (consider selects). It's a separate issue, and would need a separate solution.

**Comment by ExE-Boss** [comment]
I&nbsp;want `:empty` to&nbsp;be&nbsp;fixed, so&nbsp;that we&nbsp;can&nbsp;finally remove one&nbsp;item from&nbsp;[the&nbsp;long, ever&nbsp;expanding, list&nbsp;of&nbsp;legacy CSS&nbsp;mistakes](https://wiki.csswg.org/ideas/mistakes).

**Comment by Sora2455** [comment]
@Dan503 Just a quick note that your use-case can be done with [`:placeholder-shown`](https://caniuse.com/#feat=css-placeholder-shown):

```
input:placeholder-shown + span {
    /* styles that move the label into the input area */
}
```

**Comment by Dan503** [comment]
@Sora2455 that only works if the input has placeholder text to show. If there is no placeholder text then it doesn't work.

**Comment by Sora2455** [comment]
@Dan503 Huh, so it does. Firefox activates the pseudo-selector if there's an empty Placeholder, though Chrome requires a space. Can't test Safari from here. See [fiddle](https://jsfiddle.net/Sora2455/gf5rduy7/).

**Comment by ExE-Boss** [comment]
@Sora2455
> @Dan503 Huh, so it does. Firefox activates the pseudo-selector if there's an empty Placeholder, though Chrome requires a space. Can't test Safari from here. See [fiddle](https://jsfiddle.net/Sora2455/gf5rduy7/).

It&nbsp;doesn’t work on&nbsp;**iOS**, so&nbsp;I&nbsp;doubt it&nbsp;works on&nbsp;desktop.

**P.S.:** How&nbsp;does **Internet&nbsp;Explorer** fare with its&nbsp;`:‑ms‑input‑placeholder` pseudo‑class?

**Comment by Sora2455** [comment]
@ExE-Boss In Edge/IE they just always stay in "floated" mode, meaning that they're still usable even though the selector is unsupported. Is that what it looks like in Safari?

`:-ms-input-placeholder` _sort_ of works, only the transition only seems to happen once a hover event happens. See [fiddle](https://jsfiddle.net/Sora2455/gf5rduy7/2/embedded/result,css,html,js) (in IE, obviously).

**Comment by Dan503** [comment]
> Firefox activates the pseudo-selector if there's an empty Placeholder, though **Chrome requires a space**.

I never thought of using an empty space in the placeholder attribute to create the effect. Nice suggestion 😊

**Comment by ExE-Boss** [comment]
> Is that what it looks like in Safari?

I&nbsp;forgot to&nbsp;specify that it&nbsp;only doesn’t&nbsp;work for&nbsp;an&nbsp;empty&nbsp;placeholder.

It&nbsp;works fine‑ish with a&nbsp;space.

**Comment by TimVevida** [comment]
@lilles It has been a while since you added the counter, any results yet?

**Comment by lilles** [comment]
2% of page loads has a white-space only fail for :empty:

https://www.chromestatus.com/metrics/feature/timeline/popularity/2624

**Comment by annevk** [comment]
@lilles okay, so that means that on pages that use `:empty`, it results in a non-match 2% of the time where it would have resulted in a match if we ignored whitespace? Per the usual breaking change numbers it seems that would be too high and mean we cannot change the definition of `:empty`, right?

**Comment by ExE-Boss** [comment]
The question is, do the pages intend for `:empty` to match white‑space, or ignore white‑space?

**Comment by Dan503** [comment]
It's totally possible that the pages are using it expecting it to activate when there is still white space in the element but this is a bug in their code.

The 2% could be seen as evidence to prove how confusing `:empty` is for developers in it's current form.

**Comment by Sora2455** [comment]
@Dan503 By that logic there was no point looking at the counter, if you were going to argue for this change with or without a negligable count.

**Comment by annevk** [comment]
To be clear, I don't really think that's a question here and I don't think it's "totally possible" as developers test their pages in browsers. That they would currently be broken in some way in all browsers seems rather absurd.

**Comment by chrishtr** [comment]
> Per the usual breaking change numbers it seems that would be too high and mean we cannot change the definition of `:empty`, right?

@annevk: right. In Chromium we use the rule of thumb of much less than 0.1% for breaking changes being an acceptable percentage for breaking changes. Anything at 2% needs to have strong evidence that it won't break sites, and also a strong justification as to why it's important enough to cause this churn for site authors.

**Comment by LeaVerou** [comment]
Since web compat was mentioned:

In the HTTPArchive data, `:empty` is used on 12% (621,005) of the desktop sites surveyed and 16.25% (964,648) of the mobile sites

**Comment by foolip** [comment]
Implementing `:empty` per spec is a proposal for Interop 2023, see https://github.com/web-platform-tests/interop/issues/180. However, the [use counter](https://chromestatus.com/metrics/feature/timeline/popularity/2624) already referenced in this thread is now at 2.5%. Careful analysis of a large-ish random sample of sites hitting the use counter would be needed to show that the risk is low enough.

**Comment by brandonmcconnell** [comment]
I think a simpler route would be to have four different selectors (two of which already exist), each clearly fulfilling a different purpose, which would also help to avoid any breaking changes for that 2.5% that @foolip (via @lilles) pointed out and also to attempt to be more future proof:

* `:empty` – no children (no whitespace allowed)
* `:blank` – empty value (for form fields and related elements)
   _being discussed here:_ https://github.com/whatwg/html/issues/8451
* `:whitespace-only` – no child element nodes or non-whitespace text nodes (whitespace allowed)
* `:text-only` – no child element nodes (text and whitespace allowed)

#### Related notes:

* `:text-only` is technically possible today (AFAICT) as a decorator using `:has()`
* `:whitespace-only` could be nearly identical in implementation to `:empty` with the slight difference of assessing trimmed contents rather than untrimmed contents

<table>
  <colgroup>
    <col span="2" />
    <col />
  </colgroup>
  <thead>
    <tr>
      <th>Pseudo selector</th>
      <th>Decorator / implementation note</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><pre>:empty</pre></td>
      <td>implementation stays as is, matches elements without any contents (whitespace will not match), spec is adjusted to exclude whitespace as allowed contents to match <code>:empty</code></td>
    </tr>
    <tr></tr>
    <tr>
      <td><pre>:blank</pre></td>
      <td>matches empty values as spec'd here and discussed here: https://github.com/whatwg/html/issues/8451</td>
    </tr>
    <tr></tr>
    <tr>
      <td><pre>:whitespace-only</pre></td>
      <td>similar to <code>:empty</code> but assesses the element's trimmed contents rather than untrimmed to ignore whitespace when evaluating for emptiness</td>
    </tr>
    <tr></tr>
    <tr>
      <td><pre>:text-only</pre></td>
      <td>decorator for <code>&:not(:has(> *))</code> to match any elements without any child elements</td>
    </tr>
  </tbody>
</table>

**Comment by romainmenke** [comment]
This might be personal, but I find `:whitespace-only` and `:text-only` confusing.
It is not clear to me that these are related to `:empty`.

Something like`:empty(allow-whitespace)` or `:empty(allow-text)` makes it clear that we are matching some definition of "empty" while allowing/ignoring certain bits.

I do agree that all four features are relevant.

**Comment by brandonmcconnell** [comment]
@romainmenke I dig that! 🙌🏼

I think `:empty()` could be a pseudo selector with a single argument for any exceptions, that being `allow-whitespace` and `allow-text`, which would also leave the door open in case any more options were ever to be conceived.

The one odd-ball out would be `:blank` which I think can stay separate, as that's rather unrelated to the empty state and has more to do with the state of a form field's value. It's a little deeper than that, but we go into that in the link I mentioned before re `:blank`.

**Comment by Fast-Nick** [comment]
The argument that :empty without including whitespace (the old spec) is confusing should be rephrased as "More than 50% of developers that have not read the documentation for :empty assumes that it includes whitespace"

...if that statement is _not_ true, then changing :empty to include whitespace would _increase_ the total amount of confusion instead of decrease it. Are we really making it better or actually worse, i.e. does anyone actually have statistics for this, or is it only "many" devs found it confusing - which doesn't prove that including whitespace would decrease confusion?

**Comment by MortezaMirmojarabian** [comment]
The `:blank` pseudo-class would significantly streamline dynamic form calculations, for example in educational contexts. It helps developers easily identify empty or non-empty input fields in CSS and JavaScript. For instance, in [this Stack Overflow post](https://stackoverflow.com/q/48111746), a teacher needs to calculate student averages based on test scores, but only for fields that have been filled (since students may skip optional tests).  

Currently, developers must rely on JavaScript event listeners to manually track input changes and count non-empty fields. This approach is cumbersome. With `:blank`, a **one-liner** like:  
```  
document.querySelectorAll("input:not(:blank)").length  
```  
could dynamically determine the divisor for averages, eliminating the need for hardcoded values and reducing complexity.  

This use case highlights the practical importance of `:blank` for forms requiring conditional logic based on user input. Its implementation would not only benefit educators but also improve form handling in surveys, assessments, and other interactive applications.  

**Proposed Action:** Prioritize the implementation and standardization of `:blank` to address real-world scenarios like this. Broader browser support would empower developers to build more robust and maintainable solutions with simple, elegant approaches.

**Comment by brandonmcconnell** [comment]
To agree with @annevk's original sentiment here, I think we need to decide on a direction for `:whitespace-only` vs. `:blank` and move forward. This is a rather crucial feature for developers. Current workarounds largely rely on build tools to strip all whitespace to permit using the current implementation of `:empty`.

I agree with @Fast-Nick's point that changing the current implementation would be more confusing, not to mention that it would introduce a breaking change ([Don’t Break The Web](https://flbrack.com/posts/2023-02-15-dont-break-the-web/)).

However, I think most developers don't have a use case for `:empty` outside of what `:whitespace-only` provides, so this would solve those cases.

Developers set up their apps working a specific way, so having them suddenly switch when already live and functional could be annoying, but if this is a risk the CSSWG is willing to run, I believe that it would be net good long-term. We'll likely be writing `:empty` for decades to come, and having one API for that feature might be best, though I've argued against that in the past.

I've set up a few apps with this `:empty` vs. with-whitespace distinction, which I would have to refactor now, but those are edge cases at best. If this is obstructing development, I think browsers should move forward in implementing this the way the CSSWG resolved to [above](https://github.com/w3c/csswg-drafts/issues/1967#issuecomment-424788796).

**Comment by njvack** [comment]
Apologies for maybe commenting in a place where I don't belong? (I was looking to see if a `:blank` selector existed in draft and wound up here.)

Is the intent for this to match any _general_ elements that have no content (eg, `<span></span>`) or interactive controls that have no content (eg, `<input value="">` and `<textarea></textarea>` but _not_ `<span></span>`)? The draft spec says it's for interactive controls but some comments in this thread seem to be implying the idea is "`:empty` except whitespace nodes don't count"

Anyhow, I don't have a strong feeling about whether we need the `:empty` only-whitespace variant but I would absolutely use the `:blank` for form controls thing to improve the usability of a data collection system where I can't change markup but _can_ inject CSS.

(As an aside: Thank you to everyone who has made CSS so much better over the years. `:has`. I cannot sing its praises enough.)

**Comment by EricBMoreno** [comment]
> Is the intent for this to match any _general_ elements that have no content (eg, `<span></span>`) or interactive controls that have no content (eg, `<input value="">` and `<textarea></textarea>` but _not_ `<span></span>`)? The draft spec says it's for interactive controls but some comments in this thread seem to be implying the idea is "`:empty` except whitespace nodes don't count"

This is my understanding, going off of [Selectors Level 4](https://drafts.csswg.org/selectors/):

If you're talking about [:blank](https://drafts.csswg.org/selectors/#blank), it is intended to specifically target elements that accept user input, but do not currently have any user-entered data (so <code>input.defaultValue</code> and <code>input.value</code> are both empty strings). Currently, I believe this is only achievable if the developer includes a placeholder and uses [:placeholder-shown](https://drafts.csswg.org/selectors/#placeholder-shown-pseudo), which is simply not possible with some elements that accept input (e.g. date inputs, file inputs, select, etc.), as they don't support the placeholder attribute.

If you're talking about [:empty](https://drafts.csswg.org/selectors/#the-empty-pseudo), it's intended to select any element with zero child nodes. Current implementations include whitespace text nodes when considering :empty, though that is out of spec. So:
<code>&lt;span&gt;&lt;/span&gt;</code> Will always get selected by :empty, but
<code>&lt;span&gt;&emsp;&emsp;&lt;/span&gt;</code>  Will not be selected by :empty with current implementations (but should be).

Blocks with only a newline will also be ignored, which is where the problem with :empty really arises in my opinion. 
<code>&lt;span&gt;
&lt;/span&gt;</code>
Using :not(:has(*)) would select the empty element above, but I can't imagine that's incredibly efficient.

Updating :empty to ignore whitespace would definitely make it more useful. I use PHP frequently in my work, and not being able to include whitespace within an element can be a nightmare for readability. The spec addresses this issue, but implementations have failed to do so thus far. I can understand this to an extent if there is a significant number of websites using it, but I don't see it often (if ever).

I do believe there is a place for :blank, and I do feel it is distinct enough from :empty that the two can coexist. I agree that :blank should be for form control elements. :empty seems to be more intended for container elements (paper is blank, a bucket is empty).

At least one issue I can foresee is any element with the contenteditable attribute, since it accepts user input. Would it be considered :blank, :empty, or possibly even both or neither?

**Comment by svgeesus** [comment]
Is this issue still in a needs-more-data holding pattern?

**Comment by ExE-Boss** [comment]
@EricBMoreno
`:not(:has(*))` would also select all elements which contain only text nodes (even non‑whitespace ones, or any other non‑`Element` nodes), in addition to matching all of `:empty`.
